### DISCUSSION TITLE
Rewrites configuration is impacting router shallow without reason

### DISCUSSION THREAD

**Discussion by gkielwasser** [vercel/next.js#37061] [Help]
Hi,

Configuring `rewrites` in `next.config.js` with  any rewrite rules like the following is changing `window.history.state.options.shallow` to `true`.
```
rewrites() {
        return [{
            source: '/____Whatever____',
            destination: '/____Whatever____',
        }]
 }
```
Leaving empty array is changing `window.history.state.options.shallow` to `undefined`.
```
rewrites() {
    return []
}
```

Why such a behavior is happening ?

Thanks

**Comment by icyJoseph** [comment]
At risk of being a bit simplistic here, you can think of `history.state` as a place where you can store data when navigating to a route, https://developer.mozilla.org/en-US/docs/Web/API/History/state, and then you can look at it from anywhere in the app.

This page describes `state` a bit more https://developer.mozilla.org/en-US/docs/Web/API/History/pushState, in the context of being used by navigation.

What data you store there is up to you. In this case, Next.js has some abstraction that has `options`, `url`, and a few other internals, that I think we should not depend on at all.

Why is it happening? it's up to Next.js internals. Why do you want to know?

If you must know, I think it is because internally a class is instantiated with a private property shallow set to `undefined`, then under certain conditions, this private property is set to either `true` or `false`, depending on how the application is configured. Then when doing navigations, the `shallow` property in options changes depending on the navigation requirements.

**Reply by gkielwasser** [reply]
First, thanks for this elaborated answer. The point this shallow option is updated during navigation helped me.
It seems I'm not investigating my real issue the right way.

**Where I started**
Basically I have 2 apps that behave differently when doing browser back in development:
One preserves the scroll position, while the other one doesn't and reset the scroll to the top.

**What I have done**
In order to identify such a behavior, I have abstracted my code as much as possible between those 2 apps. 
After putting away all not identical code, the last thing I found to investigate was the `next.config.js` :)

**What is really happening**
In fact both of my app, after at least one navigation, behave exactly the same. Each time I do browser back it scroll top and it do not preserve the scrolling position.

I have also seen this is only happening in development. I cannot reproduce it on production where the scroll is well preserved after doing browser back.

**Reply by icyJoseph** [reply]
Ah, that... perhaps you ought to read thorugh https://github.com/vercel/next.js/issues/20951

For navigation you control you can disable the scrolling, but for back/forward, next is supposed to let the browser handle back/forward events, but I think there might be some ongoing issue with that? People claim this works for them, https://gist.github.com/claus/992a5596d6532ac91b24abe24e10ae81, since you have a reproduction case, maybe you could give it a go?
